### ISSUE TITLE
Update Default / Custom Invite Behavior

### ISSUE THREAD

**Issue by dangrous** [Expensify/App#51096]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___
cc @rojiphil 

A pretty quick one here - we want to update the behavior for the Workspace Invite Message Page. Namely:

As long as the RHP stays open, keep the messages as edited (even if it is set to empty). And save the edited message as invite message draft only on an actual invite. Once the RHP closes without an actual invitation, discard the edited message. And when the RHP is shown again, show either the previously saved invite message draft or show the default message if no custom message is set.

Currently, we reset to the default value whenever the invite message page is shown and the message is empty. This is irrespective of whether the RHP stays open or closed.

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

**Comment by Shahidullah-Muffakir** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
we are not saving the welcome message as draft when user send the invitation, it saved every time user edit the message.
### What is the root cause of that problem?
New feature
### What changes do you think we should make in order to solve the problem?
I think the implemnetion is very straight forward
1. don't update the message draft when user is changing the message, so remove this:
https://github.com/Expensify/App/blob/41ed362304335de00d3354f04fe9ab0150d99003/src/pages/workspace/WorkspaceInviteMessagePage.tsx#L194-L197
remove the:
`                                debouncedSaveDraft(text);
`
2. and only save it as draft when user, send invite so removed this:
`  debouncedSaveDraft(null);
`from **sendInvitation** function

**Comment by huult** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Update Default / Custom Workspace Invite Behavior

### What is the root cause of that problem?

New feature

### What changes do you think we should make in order to solve the problem?

We should save draft when sent invite:
```diff
//.src/pages/workspace/WorkspaceInviteMessagePage.tsx#105
-  debouncedSaveDraft(null);
+  debouncedSaveDraft(welcomeNote ?? '');
```

And we will clear the draft when the RBP is closed if no invite is sent
```diff
//.src/pages/workspace/WorkspaceInvitePage.tsx#L80
+ const preEmployeeList = usePrevious(policy?.employeeList);
useEffect(() => {
        return () => {
+            if (Object.keys(preEmployeeList).length === Object.keys(policy?.employeeList).length) {
                Policy.setWorkspaceInviteMessageDraft(route.params.policyID, '');
+            }
        };
     
    }, [route.params.policyID, policy]);
```

**Comment by rojiphil** [comment]
@dangrous Don't we have to put this issue on hold until #49996 is done?
Also, can I be the C+ here since I have the context? I hope @aimane-chnaif does not mind.

**Comment by dangrous** [comment]
ah yep forgot to hold it!

**Comment by ChavdaSachin** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Update Default / Custom Workspace Invite Behavior

### What is the root cause of that problem?
Behavior change on demand

### What changes do you think we should make in order to solve the problem?
This is on simpler side.....
Just reset the `workspaceInviteMessageDraft` when we render `inviteMembersPage` for the first time we already have `firstRenderRef` on `WorkspaceInvitePage` for similar applications, so this will trigger a reset each time user closes RHP and comes back to `inviteMembers` page.
```js
    Policy.setWorkspaceInviteMessageDraft(route.params.policyID, 
        Parser.htmlToMarkdown(policy?.description ?? '') ||
        translate('workspace.common.welcomeNote', {
            workspaceName: policy?.name ?? '',
        })
    );
```
And boom....
Now we can get rid of tons of mechanism related to welcomeNote and everything will work perfectly.
For simplicity of proposal I am rather posting link to my test branch instead of posting all code diffs.
[Test Branch](https://github.com/ChavdaSachin/App/tree/fix-51096/Invite-Members-Welcome-Note)

### What alternative solutions did you explore? (Optional)

**Reminder:** Please use plain English, be brief and avoid jargon. Feel free to use images, charts or pseudo-code if necessary. Do not post large multi-line diffs or write walls of text. Do not create PRs unless you have been hired for this job.

**Comment by rojiphil** [comment]
Note: We are still holding on to the dependent issue as there is a follow-up PR. Once it is done, we will pick this up.

**Comment by rojiphil** [comment]
Still on hold

**Comment by anmurali** [comment]
Held

**Comment by rojiphil** [comment]
Thanks for all your proposals. 
@Krishna2323 [proposal](https://github.com/Expensify/App/issues/51096#issuecomment-2452926581) LGTM as using form `shouldSaveDraft` seems better than managing another custom entry in Onyx to implement this feature.

However, I am not sure if we have covered all the use cases here although the test video in the proposal seems to be addressing Tests 3, 4 and 5 as mentioned in the test steps below.
Would all Tests (including Test 1 and Test 2) work in your proposal?

Precondition: No preexisting saved invite message draft should be set for the workspace.
Test Steps:
1. Go to Account Settings -> Workspaces
2. Select a workspace and go to Members->Invite member
3. Select a member to invite and click on Next to go to the invite message view. This will show the default welcome message.
4. Delete the default welcome message text so that the message text is now empty.
5. Navigate back to member selection page and click `Next` again to go to the invite message view
6. **Verify that the message text remains empty (Test 1)**
7. Dismiss the RHP either by navigating back twice or by clicking outside of the RHP
8. Click on `Invite member` to navigate to member selection page, select a member, and click on Next to go to the invite message view. 
9. **Verify that the default welcome message is displayed (Test 2).**
10. Edit the welcome message text with a custom message.
11. Navigate back to member selection page and click `Next` again to go to the invite message view
12. **Verify that the the custom message is displayed (Test 3)**
13. Dismiss the RHP either by navigating back twice or by clicking outside of the RHP
14. Click on `Invite member` to navigate to member selection page, select a member, and click on Next to go to the invite message view.
15. **Verify that the default welcome message is displayed (Test 4).**
16. Edit the welcome message text with a custom message.
17. Click `Invite` so that invitation is sent to the member and RHP is dismissed.
18. Click on `Invite member` to navigate to member selection page, select a member, and click on Next to go to the invite message view.
19. **Verify that the custom message set in Step 16 is displayed (Test 5).**

**Comment by rojiphil** [comment]
@anmurali I think we can take this out of hold now since the dependent PRs are all merged

**Comment by huult** [comment]
@rojiphil Could you take a look at [my proposal](https://github.com/Expensify/App/issues/51096#issuecomment-2422915303)? My proposal is the second one listed, and it suggests using a different key to save the draft message, which aligns with the intended behavior of this ticket.

**Comment by rojiphil** [comment]
> Could you take a look at [my proposal](https://github.com/Expensify/App/issues/51096#issuecomment-2422915303)? My proposal is the second one listed, and it suggests using a different key to save the draft message, which aligns with the intended behavior of this ticket.

> using form `shouldSaveDraft` seems better than managing another custom entry in Onyx to implement this feature.

@huult I understand your proposal. However, as mentioned above, using the already existing `shouldSaveDraft` is better than using another custom entry in Onyx i.e using a different key. This is the reason to prefer the other proposal. However, if the other proposal needs another key in Onyx to implement the feature I will consider your proposal again.

**Comment by rojiphil** [comment]
> I don't think we should show the default welcome message after inviting users with an empty message. In my opinion, we should show an empty input to keep it consistent with step 5.

If we do not show the default message again, there must be a way to bring it back when needed. Instead, I think it is better to show the default message in such a case. Anyway, let us wait for @dannymcclain to chime in.

**Comment by huult** [comment]
@rojiphil In my proposal, add a new key to handle this case:

1. Invite a member.
2. Edit the welcome message.
3. Click the "Invite" button.
4. Invite a new member.
5. Select email.
6. Edit the welcome message without clicking the "Invite" button.
7. Go back to select "Add More Users."
8. Open the welcome message again.
**9. Verify that the welcome message edited in Step 6 is still retained.**
10. Close the invite modal without clicking the "Invite" button.
11. Open the invite modal again.
**12. Verify that the welcome message from Step 2 is displayed.**

That's why I added a new key to handle this ticket."

**Comment by dannymcclain** [comment]
>If we do not show the default message again, there must be a way to bring it back when needed

This is a good point. If we don't "reset" it after an empty invite message being sent, is there _any_ way to get the default message back? I kinda think this behavior is fine/potentially even desirable.

@Expensify/design do you all have any thoughts here?

**Comment by shawnborton** [comment]
Yeah part of me thinks if you decided to get rid of the default message, then that's your choice and we should respect that. So I think I would be fine not having any way to bring it back.

**Comment by dannymcclain** [comment]
Cool, I don't feel super strongly so I'm fine with that!

**Comment by rojiphil** [comment]
> So I think I would be fine not having any way to bring it back.

Sure then. Let's do it this way. I have updated the test steps.

**Comment by rojiphil** [comment]
@Krishna2323 [proposal](https://github.com/Expensify/App/issues/51096#issuecomment-2452926581) LGTM as leveraging the existing support of form `shouldSaveDraft` is better than managing another custom entry in Onyx to implement this feature.
🎀👀🎀 C+ reviewed

**Comment by anmurali** [comment]
@jasperhuangg waiting on your review!

**Comment by anmurali** [comment]
Rotating to get another BZ assigned while OOO

**Comment by jasperhuangg** [comment]
Apologies for the delay, reviewed the proposal and it makes sense to me, thanks for the bump!

**Comment by rojiphil** [comment]
@garrettmknight This issue is an improvement and not a bug. As such, I do not think a BZ checklist is needed here. 

But, adding the following test case to the database would be ideal  to avoid any regressions:

Precondition: No preexisting saved invite message draft should be set for the workspace.
1. Go to Account Settings -> Workspaces
2. Select a workspace and go to Members->Invite member
3. Select a member to invite and click on Next to go to the invite message view. This will show the default welcome message.
4. Delete the default welcome message text so that the message text is now empty.
5. Navigate back to member selection page and click Next again to go to the invite message view
6. Verify that the message text remains empty (Test 1)
7. Dismiss the RHP either by navigating back twice or by clicking outside of the RHP
8. Click on Invite member to navigate to member selection page, select a member, and click on Next to go to the invite message view.
9. Verify that the default welcome message is displayed (Test 2).
10. Edit the welcome message text with a custom message.
11. Navigate back to member selection page and click Next again to go to the invite message view
12. Verify that the the custom message is displayed (Test 3)
13. Dismiss the RHP either by navigating back twice or by clicking outside of the RHP
14. Click on Invite member to navigate to member selection page, select a member, and click on Next to go to the invite message view.
15. Verify that the default welcome message is displayed (Test 4).
16. Edit the welcome message text with a custom message.
17. Click Invite so that invitation is sent to the member and RHP is dismissed.
18. Click on Invite member to navigate to member selection page, select a member, and click on Next to go to the invite message view.
19. Verify that the custom message set in Step 16 is displayed (Test 5).
20. Delete the custom message text so that the message text is now empty.
21. Click Invite so that invitation is sent to the member and RHP is dismissed.
22. Click on Invite member to navigate to member selection page, select a member, and click on Next to go to the invite message view.
23. Verify that the message text remains empty (Test 6).

**Comment by garrettmknight** [comment]
@rojiphil new offer out to you.

**Comment by garrettmknight** [comment]
@Krishna2323 you've been paid!

**Comment by rojiphil** [comment]
> @rojiphil new offer out to you.

@garrettmknight Accepted the offer. Thanks.

**Comment by garrettmknight** [comment]
Paid out, closing!
